### PR TITLE
Do not pack tuples

### PR DESCRIPTION
**Pull request by fbs**
Fixes: #1502 and #1501

### CODE REVIEW

**Comment by fbs** [comment]
I think this works but I want to add a few more runtime tests just to be sure.

File: src/ast/codegen_llvm.cpp
Comment (fbs) [inline src/ast/codegen_llvm.cpp:1645]: Offset mismatch (wrong vars here too), shouldn't be fatal as the offset is only relevant when printing. So we can just fix the struct. As long as the size matches it's ok

**Comment by fbs** [comment]
> I think this works but I want to add a few more runtime tests just to be sure.

Looks like sizing is off for large tuples due to alignment:

```
sudo bpftrace -e 'BEGIN { print(((int8)1,"abc",3, (int32) 1, (uint64)10, (int8)10, (int16)2)); exit(); }'
tuple {
   0 | int8
   1 | string[64]
  65 | __pad[7]
  72 | int64
  80 | int32
  84 | __pad[4]
  88 | unsigned int64
  96 | int8
  97 | __pad[1]
  98 | int16
} sizeof: [100]
FATAL: BUG: Struct size mismatch: 100 != 104
```

I'll add some more debugging tools to this so we can troubleshoot it.

**Comment by danobi** [comment]
First commit LGTM, you're still working on second commit right? Feel free to split up this PR into 2 PRs if you wanna land the first commit.

**Comment by fbs** [comment]
I think I got the struct alignment thing working correctly now. Will do some more testing tomorrow.

**Comment by fbs** [comment]
I've added a few weird cases to get some extra test coverage. In this case more == better imo so im open to suggestions.

Should be ready for review now.

**Comment by mmisono** [comment]
Something goes wrong for tuple in tuple.

```
% sudo ./src/bpftrace -e 'BEGIN{ @ = ((int8)1, ((int8)1, (int8)1)); }'
FATAL: BUG: Struct size mismatch: 4 != 3
zsh: abort      sudo ./src/bpftrace -e 'BEGIN{ @ = ((int8)1, ((int8)1, (int8)1)); }'
```

It seems the size of tuple is calculated as 1 (the size of first) + 2 (that of the second).

Another example:

```
% sudo ./src/bpftrace -e 'BEGIN{ @ = (1, 2, (int8)3, ((int8)1, (int32)-1)); }'
DEBUG: [/home/ubuntu/work/bpftrace/bpftrace/src/ast/codegen_llvm.cpp:1648]
Struct offset mismatch for: (int8,int32)(3): (llvm) 20 != 24
Attaching 1 probe...
^C

@: (1, 2, 3, (-1, 0))
```

**Comment by fbs** [comment]
Ah good catch, thanks

**Comment by fbs** [comment]
I've updated it to store the required alignment for the tuple into the tuple object which should resolve all issues :)

File: src/bpftrace.cpp
Comment (fbs) [inline src/bpftrace.cpp:1379]: this was the only way I could get lgtm to shut up, the `// lgtm` below doesn't seeem to work

File: src/ast/codegen_llvm.cpp
Comment (mmisono) [inline src/ast/codegen_llvm.cpp:1637]: nit: `layout` is more appropriate? (since this function could be used with other than tuple)

**Review by mmisono — APPROVED** [review]
I confirmed that the problems I encountered are all fixed, thanks! Just some minor comments.

File: src/ast/codegen_llvm.cpp
Comment (fbs) [reply src/ast/codegen_llvm.cpp:1647]: I don't think we should emit warnings for things that the user cannot action on. It's a bug, but we can recover from it and noone will notice it.

Debug isn't ideal either, but atleast it's not in the users face

File: src/types.h
Comment (fbs) [reply src/types.h:350]: I think we can, I'll give it a go :)

File: src/ast/codegen_llvm.cpp
Comment (mmisono) [reply src/ast/codegen_llvm.cpp:1647]: ah, fair point.

**Pull request merged**
